src/particles/pusher/PlasmaParticleAdvance.cpp

```diff
@@ -148,56 +116,190 @@ AdvancePlasmaParticles (PlasmaParticleContainer& plasma, const Fields & fields,
-                                           Fx1[ip], Fy1[ip], Fux1[ip], Fuy1[ip], Fpsi1[ip],
-                                           Fx2[ip], Fy2[ip], Fux2[ip], Fuy2[ip], Fpsi2[ip],
-                                           Fx3[ip], Fy3[ip], Fux3[ip], Fuy3[ip], Fpsi3[ip],
-                                           Fx4[ip], Fy4[ip], Fux4[ip], Fuy4[ip], Fpsi4[ip],
-                                           Fx5[ip], Fy5[ip], Fux5[ip], Fuy5[ip], Fpsi5[ip],
-                                           dz, temp_slice, ip, SetPosition, enforceBC );
+                    amrex::Real q_mass_clight_ratio = charge_mass_clight_ratio;
+                    if (ion_lev) {
+                        q_mass_clight_ratio *= ion_lev[ip];
                     }
-                    return;
+                    Bxp *= clight;
+                    Byp *= clight;
+                    Aabssqp *= 0.5_rt; // TODO: fix units of Aabssqp
+                    AabssqDxp *= 0.25_rt * me_clight_mass_ratio;
+                    AabssqDyp *= 0.25_rt * me_clight_mass_ratio;
+
+#ifndef HIPACE_USE_AB5_PUSH
+
+                    constexpr int nsub = 4;
+                    const amrex::Real sdz = dz/nsub;
+
+                    amrex::Real ux = ux_half_step[ip];
+                    amrex::Real uy = uy_half_step[ip];
+                    amrex::Real psi = psi_half_step[ip];
+
+                    for (int isub=0; isub<nsub; ++isub) {
+
+                        const amrex::Real psi_inv = 1._rt/psi;
+
+                        auto [dz_ux, dz_uy, dz_psi] = PlasmaMomentumPush(
+                            ux, uy, psi_inv, ExmByp, EypBxp, Ezp, Bxp, Byp, Bzp,
+                            Aabssqp, AabssqDxp, AabssqDyp, clight_inv, q_mass_clight_ratio);
+
+                        const DualNumber ux_dual{ux, dz_ux};
+                        const DualNumber uy_dual{uy, dz_uy};
+                        const DualNumber psi_inv_dual{psi_inv, -psi_inv*psi_inv*dz_psi};
+
+                        auto [dz_ux_dual, dz_uy_dual, dz_psi_dual] = PlasmaMomentumPush(
+                            ux_dual, uy_dual, psi_inv_dual, ExmByp, EypBxp, Ezp, Bxp, Byp, Bzp,
+                            Aabssqp, AabssqDxp, AabssqDyp, clight_inv, q_mass_clight_ratio);
+
+                        ux += sdz*dz_ux + 0.5_rt*sdz*sdz*dz_ux_dual.epsilon;
+                        uy += sdz*dz_uy + 0.5_rt*sdz*sdz*dz_uy_dual.epsilon;
+                        psi += sdz*dz_psi + 0.5_rt*sdz*sdz*dz_psi_dual.epsilon;
+
+                    }
+
+                    xp += dz*clight_inv*(ux * (1._rt/psi));
+                    yp += dz*clight_inv*(uy * (1._rt/psi));
+
+                    if (setPositionEnforceBC(ip, xp, yp)) return;
+
+                    if (!temp_slice) {
+                        ux_half_step[ip] = ux;
+                        uy_half_step[ip] = uy;
+                        psi_half_step[ip] = psi;
+                        x_prev[ip] = xp;
+                        y_prev[ip] = yp;
+                    }
```

Comment:
Could you add a comment on this handling? I think we discussed it offline, but it would be nice to have it explained in the code somehow.
```suggestion
                    if (!temp_slice) {
                        ux_half_step[ip] = ux;
                        uy_half_step[ip] = uy;
                        psi_half_step[ip] = psi;
                        x_prev[ip] = xp;
                        y_prev[ip] = yp;
                    }
```